File: src/ozw/tzemt400.cxx
/*
 * Author: Jon Trulson <jtrulson@ics.com>
 * Copyright (c) 2016 Intel Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the The MIT License which is available at
 * https://opensource.org/licenses/MIT.
 *
 * SPDX-License-Identifier: MIT
 */

#include <unistd.h>
#include <iostream>
#include <stdexcept>

#include "tzemt400.hpp"

using namespace upm;
using namespace std;

// conversion from fahrenheit to celsius and back

static float f2c(float f)
{
  return ((f - 32.0) / (9.0 / 5.0));
}

static float c2f(float c)
{
  return (c * (9.0 / 5.0) + 32.0);
}

TZEMT400::TZEMT400(int nodeID) :
  ozwInterface(nodeID)
{
}

TZEMT400::~TZEMT400()
{
}

float TZEMT400::getTemperature(bool fahrenheit)
{
  float temp = m_instance->getValueAsFloat(m_nodeID, INDEX_Temperature);
  uint8_t isFahrenheit = m_instance->getValueAsByte(m_nodeID, INDEX_FC_Units);

  if (fahrenheit)
    {
      if (isFahrenheit)
        return temp;
      else
        return c2f(temp);
    }
  else
    {
      if (isFahrenheit)
        return f2c(temp);
      else
        return temp;
    }
}

string TZEMT400::getMode()
{
  return m_instance->getValueAsString(m_nodeID, INDEX_Mode);
}

string TZEMT400::getOperatingState()
{
  return m_instance->getValueAsString(m_nodeID, INDEX_Operating_State);
}

float TZEMT400::getHeatingPointTemperature(bool fahrenheit)
{
  float temp = m_instance->getValueAsFloat(m_nodeID, INDEX_Heating_Point);
  uint8_t isFahrenheit = m_instance->getValueAsByte(m_nodeID, INDEX_FC_Units);

  if (fahrenheit)
    {
      if (isFahrenheit)
        return temp;
      else
        return c2f(temp);
    }
  else
    {
      if (isFahrenheit)
        return f2c(temp);
      else
        return temp;
    }
}

float TZEMT400::getCoolingPointTemperature(bool fahrenheit)
{
  float temp = m_instance->getValueAsFloat(m_nodeID, INDEX_Cooling_Point);
  uint8_t isFahrenheit = m_instance->getValueAsByte(m_nodeID, INDEX_FC_Units);

  if (fahrenheit)
    {
      if (isFahrenheit)
        return temp;
      else
        return c2f(temp);
    }
  else
    {
      if (isFahrenheit)
        return f2c(temp);
      else
        return temp;
    }
}

string TZEMT400::getFanMode()
{
  return m_instance->getValueAsString(m_nodeID, INDEX_Fan_Mode);
}

string TZEMT400::getFanState()
{
  return m_instance->getValueAsString(m_nodeID, INDEX_Fan_State);
}

void TZEMT400::update()
{
  m_instance->refreshValue(m_nodeID, INDEX_Temperature);
  m_instance->refreshValue(m_nodeID, INDEX_Mode);
  m_instance->refreshValue(m_nodeID, INDEX_Operating_State);
  m_instance->refreshValue(m_nodeID, INDEX_Heating_Point);
  m_instance->refreshValue(m_nodeID, INDEX_Cooling_Point);
  m_instance->refreshValue(m_nodeID, INDEX_Fan_Mode);
  m_instance->refreshValue(m_nodeID, INDEX_Fan_State);
  m_instance->refreshValue(m_nodeID, INDEX_FC_Units);
}

File: examples/javascript/tzemt400.js
/*
 * Author: Jon Trulson <jtrulson@ics.com>
 * Copyright (c) 2016 Intel Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the The MIT License which is available at
 * https://opensource.org/licenses/MIT.
 *
 * SPDX-License-Identifier: MIT
 */


var sensorObj = require('jsupm_ozw');

/************** Main code **************/
var defaultDev = "/dev/ttyACM0";

// if an argument was specified, use it as the device instead
if (process.argv.length > 2)
{
    defaultDev = process.argv[2];
}
console.log("Using device " + defaultDev);

// Instantiate a TZEMT400 instance, on device node 13.  You will
// almost certainly need to change this to reflect your own network.
// Use the ozwdump example to see what nodes are available.
var sensor = new sensorObj.TZEMT400(13);

// The first thing to do is create options, then lock them when done.
sensor.optionsCreate();
sensor.optionsLock();

// Next, initialize it.
console.log("Initializing, this may take awhile depending on your ZWave network");

sensor.init(defaultDev);
console.log("Initialization complete");

console.log("Querying data...");

setInterval(function()
{
    sensor.update();

    console.log("Temperature: "
                + sensor.getTemperature()
                + " C / "
                + sensor.getTemperature(true)
                + " F");

    console.log("Mode: "
                + sensor.getMode());

    console.log("Operating State: "
                + sensor.getOperatingState());

    console.log("Heating Point: "
                + sensor.getHeatingPointTemperature()
                + " C / "
                + sensor.getHeatingPointTemperature(true)
                + " F");

    console.log("Cooling Point: "
                + sensor.getCoolingPointTemperature()
                + " C / "
                + sensor.getCoolingPointTemperature(true)
                + " F");

    console.log("Fan Mode: "
                + sensor.getFanMode());

    console.log("Fan State: "
                + sensor.getFanState());

    console.log();
}, 5000);


// exit on ^C
process.on('SIGINT', function()
{
    sensor = null;
    sensorObj.cleanUp();
    sensorObj = null;
    console.log("Exiting.");
    process.exit(0);
});
